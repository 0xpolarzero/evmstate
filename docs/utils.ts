import { isHex } from "tevm";
import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const maxBytesLength = 64;
export const stringify = (value: any, space: number = 2) => {
  return JSON.stringify(
    value,
    (_, v) => {
      if (typeof v === "bigint") return `__bigint__${v.toString()}`;
      if (isHex(v) && (v.length - 2) / 2 > maxBytesLength)
        return `__long_hex__${v.slice(0, maxBytesLength + 2)}...${v.slice(-maxBytesLength)}`;
      return v;
    },
    space,
  );
};
