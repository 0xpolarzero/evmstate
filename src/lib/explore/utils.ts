import { Hex, toHex } from "tevm";
import { ByteArray, ToHexParameters } from "viem";

export const max = <T extends bigint | number>(...nums: Array<T>): T =>
  // @ts-expect-error bigint/number
  nums.reduce((max, num) => (num > max ? num : max), -Infinity);

/**
 * Converts a value to a hex string and ensures it has an even number of characters (full bytes)
 *
 * @param value The value to convert to hex
 * @param options Optional configuration options passed to tevm's toHex
 * @returns A hex string with full bytes
 */
export const toHexFullBytes = (value: string | number | bigint | boolean | ByteArray, opts?: ToHexParameters): Hex => {
  // Use tevm's toHex function
  const hex = toHex(value, opts);

  // If the hex string has an odd number of characters after the 0x prefix,
  // pad it with a leading zero to make it even
  if ((hex.length - 2) % 2 !== 0) return ("0x0" + hex.slice(2)) as Hex;
  return hex;
};
