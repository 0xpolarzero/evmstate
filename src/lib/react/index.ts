export { TracerProvider, useTracer } from "./lib.jsx";
