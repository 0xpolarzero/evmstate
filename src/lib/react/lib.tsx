import React, { createContext, type ReactNode, useContext, useMemo } from "react";

import { type ExploreStorageConfig } from "@/lib/explore/config.js";
import { Tracer } from "@/lib/trace/index.js";
import type { TraceStateBaseOptions } from "@/lib/trace/types.js";

// Define the context type
type TracerContextType = Tracer | null;

// Create the context with a default null value
export const TracerContext = createContext<TracerContextType>(null);

interface TracerProviderProps extends TraceStateBaseOptions {
  config?: ExploreStorageConfig;
  children: ReactNode;
}

/**
 * Provides a Tracer instance to its children components via context. It initializes the Tracer with the provided
 * options.
 */
export const TracerProvider: React.FC<TracerProviderProps> = ({ children, ...options }) => {
  // Memoize the Tracer instance to avoid recreating it on every render unless the options change.
  const tracer = useMemo(
    () => new Tracer(options),
    [options.client, options.rpcUrl, options.common, options.explorers, options.config],
  );

  return <TracerContext.Provider value={tracer}>{children}</TracerContext.Provider>;
};

/**
 * Custom hook to access the Tracer instance provided by TracerProvider.
 *
 * @returns The Tracer instance.
 * @throws Error if used outside of a TracerProvider.
 */
export const useTracer = (): Tracer => {
  const context = useContext(TracerContext);
  if (context === null) {
    throw new Error("useTracer must be used within a TracerProvider");
  }
  return context;
};
