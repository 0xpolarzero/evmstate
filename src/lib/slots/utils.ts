import { Address, decodeAbiParameters, encodeAbiParameters, Hex, isHex, toHex } from "tevm";
import { abi } from "@shazow/whatsabi";
import { AbiType, AbiTypeToPrimitiveType } from "abitype";
import { padHex } from "viem";

import { MappingKey } from "@/lib/slots/types";

/* -------------------------------------------------------------------------- */
/*                                  MAPPINGS                                  */
/* -------------------------------------------------------------------------- */

/** Extract values from a transaction trace that might be used as mapping keys */
// TODO: refactor to navigate the stack trace and identify both potential keys and array indexes
/**
 * Typical access patterns:
 *
 * Mapping
 *
 * - KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key, uint256(mappingSlot)))
 * - -> key is the input to the hash so we might even be able to retrieve the slot and compare directly
 * - KECCAK256 → KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key2, keccak256(abi.encode(key1, uint256(mappingSlot)))))
 * - -> same here we get each key from each keccak input
 *
 * Array
 *
 * - KECCAK256 → ADD → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(uint256(arraySlot)) + index
 *
 * As reference:
 *
 * - Struct in array: KECCAK256 → ADD → ADD → SLOAD/SSTORE Slot calculation: (keccak256(uint256(arraySlot)) + index) +
 *   memberOffset
 * - Mapping to Struct with Mapping: KECCAK256 → ADD → KECCAK256 → SLOAD/SSTORE Slot calculation:
 *   keccak256(abi.encode(key2, (keccak256(abi.encode(key1, uint256(mappingSlot))) + memberOffset)))
 */
export const extractPotentialKeys = (
  trace: {
    uniqueStackValues?: Array<string>;
    relevantOps?: Array<{
      op: string;
      stack: Array<string>;
    }>;
  },
  addresses: Array<Address>,
  abiFunctions: Array<abi.ABIFunction>,
  txData?: Hex,
): MappingKey[] => {
  const keys: MappingKey[] = [];

  // Add touched addresses
  addresses.forEach((address) => {
    keys.push({
      hex: padHex(address, { size: 32 }),
      decoded: address,
      type: "address",
    });
  });

  // Extract parameters from transaction data
  if (txData && txData.length > 10) {
    const selector = txData.slice(0, 10);
    const inputs = abiFunctions.find((fn) => fn.selector === selector)?.inputs;

    if (inputs) {
      // Decode function inputs
      const params = decodeAbiParameters(inputs, `0x${txData.slice(10)}`);

      params.forEach((param, index) => {
        // If it's an array, add each element as a key
        if (Array.isArray(param)) {
          param.forEach((p) => {
            const type = inputs[index].type.replace("[]", "") as AbiType;

            if (type && type !== "string" && type !== "bytes") {
              keys.push({
                hex: padHex(encodeAbiParameters([{ type }], [p]), { size: 32 }),
                decoded: p as AbiTypeToPrimitiveType<typeof type>,
                type,
              });
            }
          });
        } else {
          // Otherwise just add the key straight up
          const type = inputs[index].type as AbiType;

          if (type && type !== "string" && type !== "bytes") {
            keys.push({
              hex: padHex(encodeAbiParameters([{ type }], [param]), { size: 32 }),
              decoded: param as AbiTypeToPrimitiveType<typeof type>,
              type,
            });
          }
        }
      });
    }
  }

  // Process stack values from the trace
  if (trace.uniqueStackValues?.length) {
    // Process unique stack values directly
    for (const stackValue of trace.uniqueStackValues) {
      keys.push({
        hex: isHex(stackValue) ? padHex(stackValue, { size: 32 }) : toHex(stackValue, { size: 32 }),
        type: undefined,
      });
    }
  }

  // Deduplicate keys
  const uniqueMap = new Map();
  // Add the new key only if it's not already in the map (and don't replace a key with a defined type)
  keys.forEach((k) => {
    if (!uniqueMap.has(k.hex) || k.type) uniqueMap.set(k.hex, k);
  });

  return Array.from(uniqueMap.values()).sort((a, b) => {
    // prefer address as it's more likely to be a key
    if (a.type === "address") return -1;
    if (b.type === "address") return 1;
    // prefer defined types
    if (a.type === undefined) return 1;
    if (b.type === undefined) return -1;

    return 0;
  });
};

/* -------------------------------------------------------------------------- */
/*                                    UTILS                                   */
/* -------------------------------------------------------------------------- */

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
