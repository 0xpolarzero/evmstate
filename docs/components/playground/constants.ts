import { EthjsAddress } from "tevm/utils";
import type { ExtractAbiFunctions } from "abitype";

import * as LAYOUTS from "@test/codegen/layouts/index.js";
import * as CONTRACTS from "@test/contracts/index.js";

export const contract = CONTRACTS["Playground"].withAddress(
  EthjsAddress.fromString("0x987C2AF139EAEaBdF8D6d3d1723C1883bEa1f2AF").toString(),
);
export const layout = LAYOUTS["Playground"];
export const callerAddress = EthjsAddress.fromString("0xCa11e40000000000000000000000000000000000");
export const localStorageKey = "EVMSTATE_PLAYGROUND_STATE";

export const functionDescriptions: Record<ExtractAbiFunctions<typeof contract.abi, "nonpayable">["name"], string> = {
  addValue: "Adds to a dynamic array",
  addUser: "Adds a user struct to a mapping and array",
  toggleUserActive: "Toggles a boolean in a struct within a mapping",
  setBalance: "Updates a simple mapping value",
  setAllowance: "Updates a nested mapping value",
  addTransaction: "Adds to a dynamic array within a mapping",
  updateBasicValues: "Updates primitive types (uint256, bool)",
  updatePackedValues: "Updates packed storage variables (uint8, uint16, uint32, bool)",
  setStringAndBytes: "Updates string and bytes storage",
  setFixedValue: "Updates a fixed-size array element",
};
