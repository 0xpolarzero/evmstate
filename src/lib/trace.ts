import { Abi, Address, CallResult, ContractFunctionName, Hex } from "tevm";

import { debug } from "@/debug";
import { createAccountDiff, intrinsicDiff, intrinsicSnapshot, storageDiff, storageSnapshot } from "@/lib/access-list";
import { extractPotentialKeys } from "@/lib/slot-engine";
import { formatLabeledStorageOp, getContracts, getStorageLayout } from "@/lib/storage-layout";
import {
  LabeledStorageRead,
  LabeledStorageWrite,
  StorageAccessTrace,
  StorageSlotInfo,
  TraceStorageAccessOptions,
  TraceStorageAccessTxParams,
} from "@/lib/types";
import { createClient /* , uniqueAddresses */, getUnifiedParams } from "@/lib/utils";

/**
 * Analyzes storage access patterns during transaction execution.
 *
 * Identifies which contract slots are read from and written to, with human-readable labels.
 *
 * Note: If you provide a Tevm client yourself, you're responsible for managing the fork's state; although default
 * mining configuration is "auto", so unless you know what you're doing, it should be working as expected intuitively.
 *
 * @example
 *   const analysis = await traceStorageAccess({
 *     from: "0x123",
 *     to: "0x456",
 *     data: "0x1234567890",
 *     client: memoryClient,
 *   });
 *
 * @param options - {@link TraceStorageAccessOptions}
 * @returns Promise<Record<Address, StorageAccessTrace>>
 */
export const traceStorageAccess = async <
  TAbi extends Abi | readonly unknown[] = Abi,
  TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
>(
  args: TraceStorageAccessOptions & TraceStorageAccessTxParams<TAbi, TFunctionName>,
): Promise<Record<Address, StorageAccessTrace>> => {
  const { client, from, to, data } = await getUnifiedParams(args);

  // Execute call on local vm with access list generation and trace
  let callResult: CallResult | undefined;
  try {
    callResult = await client.tevmCall({
      from,
      to,
      data,
      skipBalance: true,
      createAccessList: true,
      createTransaction: true,
      createTrace: true,
    });

    if (callResult.errors) {
      debug(`EVM exception during call: ${callResult.errors.map((err) => err.message).join(", ")}`);
      throw new Error(callResult.errors.map((err) => err.message).join(", "));
    }
  } catch (err) {
    debug(`Failed to execute call: ${err}`);
    throw err;
  }

  // Debug log showing the trace size and unique stack values
  debug(
    `Trace contains ${callResult.trace?.structLogs.length} steps and ${[...new Set(callResult.trace?.structLogs.flatMap((log) => log.stack))].length} unique stack values`,
  );

  // Get all relevant addresses (contract addresses + sender + target + any created contracts)
  // const addresses = uniqueAddresses([
  //   ...(Object.keys(callResult.accessList ?? {}) as Address[]),
  //   from,
  //   to,
  //   ...((callResult.createdAddresses ?? []) as Address[]),
  // ]);
  // TODO: research to make sure this really includes all relevant addresses but it should (all accounts touched by the tx)
  // currently enabled with createAccessList: true
  const addresses = Object.values(callResult.preimages ?? {}).filter(
    (address) => address !== "0x0000000000000000000000000000000000000000",
  );
  debug(`${addresses.length} accounts touched during the transaction`);

  // Get the storage and account values before the transaction is mined
  const storagePreTx = await storageSnapshot(client, callResult.accessList ?? {});
  const intrinsicsPreTx = await intrinsicSnapshot(client, addresses);

  // Mine the pending transaction to get post-state values
  await client.tevmMine();

  // const debugCall = await client.request({
  //   method: "debug_traceTransaction",
  //   params: [
  //     callResult.txHash,
  //     {
  //       tracer: "prestateTracer",
  //       tracerConfig: {
  //         diffMode: true,
  //       },
  //     },
  //   ],
  // });
  // console.log(debugCall);

  // Get values after the transaction has been included
  const storagePostTx = await storageSnapshot(client, callResult.accessList ?? {});
  const intrinsicsPostTx = await intrinsicSnapshot(client, addresses);

  // Get contracts' storage layouts
  const filteredContracts = addresses.filter((address) => {
    const preTxStorage = storagePreTx[address];
    const postTxStorage = storagePostTx[address];
    return preTxStorage && postTxStorage;
  });

  const contractsInfo = await getContracts({ client, addresses: filteredContracts, explorers: args.explorers });

  // Map to store storage layouts per contract
  const storageLayouts: Record<Address, Array<StorageSlotInfo>> = {};

  // Get storage layouts for each contract
  await Promise.all(
    Object.entries(contractsInfo).map(async ([address, contract]) => {
      const layoutResult = await getStorageLayout({ ...contract, address: address as Address });
      if (layoutResult?.labeledSlots) {
        storageLayouts[address as Address] = layoutResult.labeledSlots;
      }
    }),
  );

  // Extract potential key/index values from the execution trace
  const traceLog = callResult.trace?.structLogs || [];

  // Create a slim version of the trace with deduplicated stack values for efficiency
  const dedupedTraceLog = {
    // Deduplicate stack values across all operations
    uniqueStackValues: [...new Set(traceLog.flatMap((log) => log.stack))],
    // Only keep storage-related operations for detailed analysis
    relevantOps: traceLog.filter((log) => ["SLOAD", "SSTORE", "SHA3"].includes(log.op)),
  };

  // Aggregate functions from all abis to be able to figure out types of args
  const abis = Object.values(contractsInfo)
    .flatMap((contract) => contract.abi)
    .filter((abi) => abi.type === "function");

  const potentialKeys = extractPotentialKeys(dedupedTraceLog, addresses, abis, data);
  debug(`Extracted ${potentialKeys.length} unique potential values from the trace`);

  // Process each address and create enhanced trace with labels
  const labeledTrace = addresses.reduce(
    (acc, address) => {
      const storage = { pre: storagePreTx[address], post: storagePostTx[address] };
      const intrinsics = { pre: intrinsicsPreTx[address], post: intrinsicsPostTx[address] };

      // Skip if this address has no relevant data
      if (!intrinsics.pre || !intrinsics.post) {
        debug(`Missing account state information for address ${address}`);
        return acc;
      }

      // For EOAs (accounts without code) we won't have storage data
      const slotsDiff =
        storage.pre && storage.post ? storageDiff(storage.pre, storage.post) : { reads: {}, writes: {} };

      // Compare account state changes
      const accountDiff = intrinsicDiff(intrinsics.pre, intrinsics.post);

      // Combine into complete diff
      const trace = createAccountDiff(slotsDiff, accountDiff);

      // Get storage layout for this contract
      const contractLayout = storageLayouts[address];

      // Create labeled reads and writes
      const labeledReads = Object.entries(trace.reads).reduce(
        (acc, [slotHex, read]) => {
          acc[slotHex as Hex] = formatLabeledStorageOp({
            op: read,
            slot: slotHex as Hex,
            contractLayout,
            potentialKeys,
          });
          return acc;
        },
        {} as Record<Hex, [LabeledStorageRead, ...LabeledStorageRead[]]>,
      );

      const labeledWrites = Object.entries(trace.writes).reduce(
        (acc, [slotHex, write]) => {
          acc[slotHex as Hex] = formatLabeledStorageOp({
            op: write,
            slot: slotHex as Hex,
            contractLayout,
            potentialKeys,
          });
          return acc;
        },
        {} as Record<Hex, [LabeledStorageWrite, ...LabeledStorageWrite[]]>,
      );

      // Return enhanced trace with labels
      acc[address] = {
        reads: labeledReads,
        writes: labeledWrites,
        intrinsic: trace.intrinsic,
      };

      return acc;
    },
    {} as Record<Address, StorageAccessTrace>,
  );

  return labeledTrace;
};

/**
 * A class that encapsulates the storage access tracing functionality.
 *
 * Allows for creating a reusable tracer with consistent configuration.
 */
export class Tracer {
  private client;
  private explorers;

  /**
   * Creates a new Tracer instance with configuration for tracing storage access.
   *
   * @param options Configuration options for the tracer
   */
  constructor(options: TraceStorageAccessOptions) {
    this.client = options.client ?? createClient(options);
    this.explorers = options.explorers;
  }

  /**
   * Traces storage access for a transaction.
   *
   * Uses the same underlying implementation as the standalone {@link traceStorageAccess} function.
   */
  async traceStorageAccess(txOptions: {
    from: Address;
    to: Address;
    data: Hex;
  }): Promise<Record<Address, StorageAccessTrace>> {
    // TODO: do we need to update the fork here? or is the "latest" blockTag enough?
    return traceStorageAccess({
      ...txOptions,
      client: this.client,
      explorers: this.explorers,
    });
  }

  // TODO: overload traceStorageAccess to accept abi, functionName, args
  // TODO: overload traceStorageAccess to accept txHash
}
