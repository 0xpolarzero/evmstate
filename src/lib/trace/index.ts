import { Abi, AbiFunction, Address, CallResult, ContractFunctionName, Hex } from "tevm";
import { SolcStorageLayout, SolcStorageLayoutTypes } from "tevm/bundler/solc";
import { AbiStateMutability, ContractFunctionArgs, toFunctionSignature, trim } from "viem";

import { debug } from "@/debug";
import { exploreStorage } from "@/lib/explore";
import { ExploreStorageConfig, parseConfig } from "@/lib/explore/config";
import { extractPotentialKeys } from "@/lib/explore/mapping";
import { intrinsicDiff, intrinsicSnapshot, storageDiff, storageSnapshot } from "@/lib/trace/access-list";
import { getContracts, getStorageLayout } from "@/lib/trace/storage-layout";
import {
  LabeledStorageAccess,
  StorageAccessTrace,
  TraceStorageAccessOptions,
  TraceStorageAccessTxParams,
  TraceStorageAccessTxWithAbi,
  TraceStorageAccessTxWithData,
  TraceStorageAccessTxWithReplay,
} from "@/lib/trace/types";
import { cleanTrace, createClient /* , uniqueAddresses */, getUnifiedParams } from "@/lib/trace/utils";

/**
 * Analyzes storage access patterns during transaction execution.
 *
 * Identifies which contract slots are read from and written to, with human-readable labels.
 *
 * Note: If you provide a Tevm client yourself, you're responsible for managing the fork's state; although default
 * mining configuration is "auto", so unless you know what you're doing, it should be working as expected intuitively.
 *
 * @example
 *   const analysis = await traceStorageAccess({
 *     from: "0x123",
 *     to: "0x456",
 *     data: "0x1234567890",
 *     client: memoryClient,
 *   });
 *
 * @param options - {@link TraceStorageAccessOptions}
 * @returns Promise<Record<Address, {@link StorageAccessTrace}>> - Storage access trace with labeled slots and labeled
 *   layout access for each touched account
 */
export const traceStorageAccess = async <
  TAbi extends Abi | readonly unknown[] = Abi,
  TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
>(
  args: TraceStorageAccessOptions & TraceStorageAccessTxParams<TAbi, TFunctionName> & { config?: ExploreStorageConfig },
): Promise<Record<Address, StorageAccessTrace>> => {
  const { client, from, to, data, value } = await getUnifiedParams(args);

  // Execute call on local vm with access list generation and trace
  let callResult: CallResult | undefined;
  try {
    callResult = await client.tevmCall({
      from,
      to,
      data,
      value,
      skipBalance: true,
      createAccessList: true,
      addToMempool: true,
      createTrace: true,
    });

    if (callResult.errors) {
      debug(`EVM exception during call: ${callResult.errors.map((err) => err.message).join(", ")}`);
      throw new Error(callResult.errors.map((err) => err.message).join(", "));
    }
  } catch (err) {
    debug(`Failed to execute call: ${err}`);
    throw err;
  }

  // Debug log showing the trace size and unique stack values
  debug(
    `Trace contains ${callResult.trace?.structLogs.length} steps and ${[...new Set(callResult.trace?.structLogs.flatMap((log) => log.stack))].length} unique stack values`,
  );

  const addresses = Object.values(callResult.preimages ?? {}).filter(
    (address) => address !== "0x0000000000000000000000000000000000000000",
  );
  debug(`${addresses.length} accounts touched during the transaction`);

  // Get the storage and account values before the transaction is mined
  const storagePreTx = await storageSnapshot(client, callResult.accessList ?? {});
  const intrinsicsPreTx = await intrinsicSnapshot(client, addresses);

  // Mine the pending transaction to get post-state values
  await client.tevmMine();

  // TODO(later): just use a diff tracer
  // const debugCall = await client.request({
  //   method: "debug_traceTransaction",
  //   params: [
  //     callResult.txHash,
  //     {
  //       tracer: "prestateTracer",
  //       tracerConfig: {
  //         diffMode: true,
  //       },
  //     },
  //   ],
  // });
  // console.log(debugCall);

  // Get values after the transaction has been included
  const storagePostTx = await storageSnapshot(client, callResult.accessList ?? {});
  const intrinsicsPostTx = await intrinsicSnapshot(client, addresses);

  // Retrieve information about the contracts for which we need the storage layout
  const contractsInfo = await getContracts({
    client,
    addresses: addresses.filter((address) => storagePreTx[address] && storagePostTx[address]),
    explorers: args.explorers,
  });

  // Map to store storage layouts adapter per contract
  const layouts: Record<Address, SolcStorageLayout> = {};

  // Get layout adapters for each contract
  await Promise.all(
    Object.entries(contractsInfo).map(async ([address, contract]) => {
      // Get storage layout adapter for this contract
      const layout = await getStorageLayout({ ...contract, address: address as Address });
      if (layout) layouts[address as Address] = layout;
    }),
  );
  // Extract potential key/index values from the execution trace
  const traceLog = callResult.trace?.structLogs || [];

  // Create a slim version of the trace with deduplicated stack values for efficiency
  const dedupedTraceLog = {
    // Deduplicate stack values across all operations
    uniqueStackValues: [...new Set(traceLog.flatMap((log) => log.stack))],
    // Only keep storage-related operations for detailed analysis
    relevantOps: traceLog.filter((log) => ["SLOAD", "SSTORE", "SHA3"].includes(log.op)),
  };

  // Aggregate functions from all abis to be able to figure out types of args
  let abis = Object.values(contractsInfo)
    .flatMap((contract) => contract.abi)
    .filter((abi) => abi.type === "function");

  // In case the tx was a contract call with the abi and it could not be fetch, add it so we can decode potential mapping keys
  if (args.abi && args.functionName) {
    const functionDef = (args.abi as Abi).find(
      (func) => func.type === "function" && func.name === args.functionName,
    ) as AbiFunction | undefined;
    // @ts-expect-error readonly/mutable types
    if (functionDef) abis.push({ ...functionDef, selector: toFunctionSignature(functionDef) });
  }

  const potentialKeys = extractPotentialKeys(dedupedTraceLog, addresses, abis, data);
  debug(`Extracted ${potentialKeys.length} unique potential values from the trace`);

  // Process each address and create enhanced trace with labels
  const LabeledStorageAccessTrace = addresses.reduce(
    (acc, address) => {
      const storage = { pre: storagePreTx[address], post: storagePostTx[address] };
      const intrinsics = { pre: intrinsicsPreTx[address], post: intrinsicsPostTx[address] };

      // Skip if this address has no relevant data
      if (!intrinsics.pre || !intrinsics.post) {
        debug(`Missing account state information for address ${address}`);
        return acc;
      }

      // For EOAs (accounts without code) we won't have storage data
      const storageTrace = storage.pre && storage.post ? storageDiff(storage.pre, storage.post) : {};

      const layout = layouts[address];
      if (!layout) {
        acc[address] = {
          storage: Object.fromEntries(
            Object.entries(storageTrace).map(([slot, { current, next }]) => [
              `slot_${slot}`,
              {
                name: `slot_${slot}`,
                trace: [
                  cleanTrace({
                    modified: next !== undefined && next !== current,
                    current: { hex: current },
                    next: { hex: next },
                    slots: [slot],
                    path: [],
                    fullExpression: `slot_${slot}`,
                    note: "Could not label this slot access because no layout was found.",
                  }),
                ],
              },
            ]),
          ),
          intrinsic: intrinsicDiff(intrinsics.pre, intrinsics.post),
        };

        return acc;
      }

      // 1. Decode using all known variables with optimized exploration
      const { withLabels, unexploredSlots } = exploreStorage(
        layout,
        storageTrace,
        potentialKeys.map((k) => k.hex),
        parseConfig(args.config),
      );

      // 1. Process results into named variables - convert all results to LabeledStorageAccess format
      const decoded = withLabels.reduce(
        (acc, result) => {
          // Retrieve existing entry or create a new one
          acc[result.name] = acc[result.name] ?? {
            name: result.name,
            type: result.type,
            kind: result.type.startsWith("mapping")
              ? "mapping"
              : result.type.endsWith("]") && result.type.match(/\[\d+\]$/)
                ? "static_array"
                : result.type.endsWith("[]")
                  ? "dynamic_array"
                  : result.type.startsWith("struct")
                    ? "struct"
                    : result.type === "bytes" || result.type === "string"
                      ? "bytes"
                      : "primitive",
            trace: [],
          };

          acc[result.name].trace.push(
            cleanTrace({
              modified: result.next !== undefined && trim(result.next.hex) !== trim(result.current.hex),
              current: result.current,
              next: result.next,
              slots: result.slots,
              path: result.path,
              fullExpression: result.fullExpression,
              note: result.note,
            }),
          );

          return acc;
        },
        {} as Record<string, LabeledStorageAccess<string, string, SolcStorageLayoutTypes>>,
      );

      // 2. Create unknown variables access traces for remaining slots
      const unknownAccess = Object.fromEntries(
        [...unexploredSlots].map((slot) => {
          const current = storageTrace[slot].current;
          const next = storageTrace[slot].next;

          return [
            `slot_${slot}`,
            {
              name: `slot_${slot}`,
              trace: [
                cleanTrace({
                  modified: next !== undefined && next !== current,
                  current: { hex: current },
                  next: { hex: next },
                  slots: [slot],
                  path: [],
                  fullExpression: `slot_${slot}`,
                  note: "Could not label this slot access.",
                }),
              ],
            },
          ];
        }),
      );

      // Return enhanced trace with labels
      acc[address] = {
        storage: { ...decoded, ...unknownAccess },
        intrinsic: intrinsicDiff(intrinsics.pre, intrinsics.post),
      };

      return acc;
    },
    {} as Record<Address, StorageAccessTrace>,
  );

  return LabeledStorageAccessTrace;
};

/**
 * A class that encapsulates the storage access tracing functionality.
 *
 * Allows for creating a reusable tracer with consistent configuration.
 */
export class Tracer {
  private options: TraceStorageAccessOptions & { config?: ExploreStorageConfig };

  /**
   * Creates a new Tracer instance with configuration for tracing storage access.
   *
   * @param options Configuration options for the tracer
   */
  constructor(options: TraceStorageAccessOptions & { config?: ExploreStorageConfig }) {
    this.options = {
      ...options,
      client: options.client ?? createClient(options),
    };

    // Bind 'this' for the traceStorageAccess method to ensure correct context
    // if it were ever destructured or passed as a callback.
    this.traceStorageAccess = this.traceStorageAccess.bind(this);
  }

  /**
   * Traces storage access for a transaction using calldata.
   *
   * Uses the same underlying implementation as the standalone {@link traceStorageAccess} function.
   */
  async traceStorageAccess(txOptions: TraceStorageAccessTxWithData): Promise<Record<Address, StorageAccessTrace>>;

  /**
   * Traces storage access for a transaction using the contract ABI.
   *
   * Uses the same underlying implementation as the standalone {@link traceStorageAccess} function.
   */
  async traceStorageAccess<
    TAbi extends Abi | readonly unknown[] = Abi,
    TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
  >(txOptions: TraceStorageAccessTxWithAbi<TAbi, TFunctionName>): Promise<Record<Address, StorageAccessTrace>>;

  /**
   * Traces storage access for a transaction using a txHash.
   *
   * Uses the same underlying implementation as the standalone {@link traceStorageAccess} function.
   */
  async traceStorageAccess(txOptions: TraceStorageAccessTxWithReplay): Promise<Record<Address, StorageAccessTrace>>;

  /**
   * Traces storage access for a transaction regardless of the parameters.
   *
   * Uses the same underlying implementation as the standalone {@link traceStorageAccess} function.
   */
  async traceStorageAccess(txOptions: TraceStorageAccessTxParams): Promise<Record<Address, StorageAccessTrace>> {
    return traceStorageAccess({
      ...this.options,
      ...txOptions,
    });
  }
}
