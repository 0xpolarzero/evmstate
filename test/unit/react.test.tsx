// @vitest-environment jsdom
import { render, screen } from "@testing-library/react";
import React, { useEffect, useState } from "react";
import { describe, expect, it } from "vitest";

import { ACCOUNTS, CONTRACTS } from "@test/constants";
import { getClient } from "@test/utils";
import { Tracer } from "@/lib/trace";
import { TracerProvider, useTracer } from "@/react";

const { StoragePacking } = CONTRACTS;
const { caller } = ACCOUNTS;

// 1. Test Component using the hook successfully
const TestComponent = () => {
  const tracer = useTracer();
  const isTracerInstance = tracer instanceof Tracer;
  return <div>Tracer instance found: {isTracerInstance ? "yes" : "no"}</div>;
};

// 2. Test Component that will cause the hook to throw (used outside provider)
const ErrorComponent = () => {
  let errorMessage = "No error";
  try {
    useTracer();
  } catch (e) {
    if (e instanceof Error) {
      errorMessage = e.message;
    } else {
      errorMessage = "Caught non-Error object";
    }
  }
  return <div>Hook Result: {errorMessage}</div>;
};

// 3. Test Component that will correctly trace storage access
const StateComponent = () => {
  const { traceState } = useTracer();
  const [trace, setTrace] = useState<string | undefined>();
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    (async () => {
      setIsLoading(true);
      setError(null);
      try {
        const trace = await traceState({
          ...StoragePacking.write.setSmallValues(1, 2, true, caller.toString()),
          from: caller.toString(),
        });
        setTrace(JSON.stringify(trace, (_, v) => (typeof v === "bigint" ? v.toString() : v), 2));
      } catch (e) {
        setError(e instanceof Error ? e.message : "An unknown error occurred");
      } finally {
        setIsLoading(false);
      }
    })();
  }, [traceState]);

  if (isLoading) return <div>Loading storage trace...</div>;
  if (error) return <div>Error: {error}</div>;

  return <div>Storage trace: {trace}</div>;
};

describe("React", () => {
  it("should provide Tracer instance via useTracer when wrapped in TracerProvider", () => {
    render(
      <TracerProvider client={getClient()}>
        <TestComponent />
      </TracerProvider>,
    );

    expect(screen.getByText("Tracer instance found: yes")).toBeDefined();
  });

  it("should throw an error when useTracer is used outside of TracerProvider", () => {
    // Render the component that attempts to use the hook without a provider
    render(<ErrorComponent />);
    expect(screen.getByText("Hook Result: useTracer must be used within a TracerProvider")).toBeDefined();
  });

  it("should correctly trace storage access via hook", async () => {
    render(
      <TracerProvider client={getClient()}>
        <StateComponent />
      </TracerProvider>,
    );

    expect(screen.getByText("Loading storage trace...")).toBeDefined();
    const resultElement = await screen.findByText(/^Storage trace: \{/i, {}, { timeout: 10000 });
    expect(JSON.parse(resultElement.textContent?.replace("Storage trace: ", "") ?? "")).toMatchSnapshot();
  });
});
