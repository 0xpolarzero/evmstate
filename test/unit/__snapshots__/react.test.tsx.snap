// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`React > should correctly trace storage access via hook 1`] = `{}`;
